### ISSUE TITLE
Docstring for twisted.trial.unittest.SynchronousTestCase.mktemp not informative enough

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6165]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| meissenPlate reported|
|-|-|
|Trac ID|trac#6165|
|Type|enhancement|
|Created|2012-10-28 17:49:15Z|
|Branch|https://github.com/twisted/twisted/tree/mktemp-api-docs-6165|

twisted.trial.unittest.SynchronousTestCase.mktemp is used to create temporary directories and files in unit tests. This is its docstring:  

```
        """
        Returns a unique name that may be used as either a temporary directory
        or filename.

        @note: you must call os.mkdir on the value returned from this method if
            you wish to use it as a directory!

        @return: C{str}
        """ 
```

Unique in what context? Could it conflict with a name outside the current working directory? Outside of a standard Twisted distribution? Or is it a magic function that magically gives universally unique names?

Generated Doc.s: http://twistedmatrix.com/documents/current/api/twisted.trial.unittest.SynchronousTestCase.html#mktemp

Source: http://twistedmatrix.com/trac/browser/tags/releases/twisted-12.2.0/twisted/trial/unittest.py#L947

Improved docstring:
```
        """
        Return a relative path that is guaranteed to be unique within the
        current working directory. Create every directory between the current 
        working directory and the last one if necessary. Do not create the 
        last directory/file.
        
        For a temporary directory call os.mkdir on the path, for a temporary
        file, just create the file (e.g. by opening the path for writing and
        then closing it).

        Trial should delete the temporary file automatically.

        @return: C{str}
        """
```

Internally it uses tempfile.mkdtemp (from the standard library), in combination with the name of the test module, test suite, and test.

It CREATES a short directory hierarchy like ".\twisted.test.test_compat\ExecfileCompatTestCase\test_execfileGlobals\fqsdxf" and returns a string like ".\twisted.test.test_compat\ExecfileCompatTestCase\test_execfileGlobals\fqsdxf\temp". The random portion is coming from tempfile.mkdtemp.

Attachments:

* [ticket_6165.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c6/9c6b5171d4c6f4a1959f21ee1c008be6edd4fa9a/ce016638b98b204ed3dc881f30f400b3b973d4f3.diff) (1354 bytes) - added by meissenPlate on 2012-10-28 17:59:06Z - Improve docstring for twisted.trial._synctest.SynchronousTestCase.mktemp.
* [ticket_6165.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c6/9c6b5171d4c6f4a1959f21ee1c008be6edd4fa9a/3e3dc1663ad79000a8b264625a60f156596d9309.diff) (1361 bytes) - added by meissenPlate on 2012-10-28 18:03:05Z - Fixed a dumb typo in the first patch's topfile entry.
* [ticket_6165_3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c6/9c6b5171d4c6f4a1959f21ee1c008be6edd4fa9a/9903b8d638d3a5dc478e2217b67bdadfdaee93c8.diff) (1364 bytes) - added by meissenPlate on 2012-12-12 17:39:36Z - Discovered misconfigured vimrc: was not auto-deleting trailing whitespace;; fixed whitespace.

<details><summary>Searchable metadata</summary>

```
trac-id__6165 6165
type__enhancement enhancement
reporter__meissenPlate meissenPlate
priority__low low
milestone__ 
branch__branches_mktemp_api_docs_6165 branches/mktemp-api-docs-6165
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__documentation documentation
time__1351446555000000 1351446555000000
changetime__1360367175000000 1360367175000000
version__None None
owner__tom_prince tom.prince
cc__jml cc__tom.prince
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

Experimentally, it appears that trial doesn't in fact delete the directory so created.

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|meissenPlate commented|
|-|-|

You are correct. (Though Trial will delete the files the next time it runs.) Also the directory isn't even created in the CWD. I don't know where I got that. It's created in twisted/_trial_temp.

Maybe the original docstring is best afterall. Just call open or mkdir on the path and run with it, don't worry about where it is or what happens to it. Probably its not worth documenting anything more specific than that.

I'm inclined to close this as wontfix and forget about it. Sorry for the trouble.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince commented|
|-|-|

Well, it is probably worth documenting that the path is relative to the current directory. And that the path returned is a fresh name in a directory (which is a subdirectory of the current directory).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

Indeed.  Thanks for working on this.  It seems like at least part of this change is desirable.  I'll see if I can split that part out.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [36876]) Branching to 'mktemp-api-docs-6165'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

How about that.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set owner to @tomprince|
|-|-|

This looks good. I'll merge it.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince set status to `closed`|
|-|-|

(In [37071]) Merge mktemp-api-docs-6165: Improve documentation of twisted.trial.unittest.SynchronousTestCase.mktemp

Author: meissenPlate, exarkun
Reviewers: exarkun, tom.prince
Fixes: [#6165](https://github.com/twisted/twisted/issues/6165)

The original documentation didn't give any direction no how it should
be use used. Explain what it returns and how to use it.
